tools/afl-fuzzer/main.c

```diff
@@ -0,0 +1,476 @@
+				dsp_run_file);
+			exit(EXIT_FAILURE);
+		}
+
+		ch = 'y';
+		if (fputc(ch, fp1) != ch) {
+			fprintf(stderr, "error: writing to dsp initializer file %s\n",
+				dsp_run_file);
+			exit(EXIT_FAILURE);
+		}
+		fclose(fp1);
+	}
+
+	// Only run fork() if ch is 'y' else don't.
+	if (ch == 'y' && !fork()) {
+
+		char *newargv[] = { "/home/sof/work/qemu/build/xtensa-softmmu/qemu-system-xtensa",
+				    "-cpu", "baytrail", "-M", "adsp_byt", "-nographic", "-kernel",
+				    "/home/sof/work/sof.git/build_byt_gcc/src/arch/xtensa/sof-byt.ri",
+				    NULL};
+		char *newenviron[] = { NULL };
+
+		printf("Executing qemu xtensa\n");
+
+		execve(newargv[0], newargv, newenviron);
+
+		/* execve() returns only on error */
+		perror("execve");
+
+		exit(EXIT_FAILURE);
+
+	} else {
+
+		// TODO: Need to make sure that child is executed first.
+		// Need to sleep only if ch = 'y' as only then DSP is restarted.
+		if (ch == 'y') {
+			sleep(5);
+
+			// In the next iteration we needn't restart DSP
+			// unless there is FW crash
+			fp1 = fopen(dsp_run_file, "w");
+
+			if (!fp1) {
+				fprintf(stderr, "error: opening dsp initializer file %s\n",
+					dsp_run_file);
+				exit(EXIT_FAILURE);
+			}
+
+			ch = 'n';
+			if (fputc(ch, fp1) != ch) {
+				fprintf(stderr, "error: writing to dsp initializer file %s\n",
+					dsp_run_file);
+				exit(EXIT_FAILURE);
+			}
+
+			// TODO: Check error conditions for this.
+			fclose(fp1);
+		}
+
+
+
+		/* init platform */
+		fprintf(stdout, "initialising platform %s\n", platform[0]->name);
+		ret = platform[0]->init(&fuzzer, platform[0], to_boot);
+
+		if (ret == ETIMEDOUT) {
+			fprintf(stderr, "error: platform %s failed to initialise\n",
+				platform_name);
+			exit(EXIT_FAILURE);
+		}
+
+		fprintf(stdout, "FW boot complete\n");
+
+		/* allocate max ipc size bytes for the msg and reply */
+		fuzzer.msg.msg_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+		fuzzer.msg.reply_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+
+		/* Open ipc message file */
+		fp = fopen(ipc_msg_file, "rb");
+		if (!fp) {
+			fprintf(stderr, "error: opening ipc msg file %s\n",
+				ipc_msg_file);
+			exit(EXIT_FAILURE);
+		}
+
+		/* ipc message file structure:
+		   header
+		   msg_size
+		   msg data
+		*/
+
+		ret = fread(&fuzzer.msg.header, sizeof(uint32_t), 1, fp);
+
+		if (ret != 1) {
+			fprintf(stderr, "error: reading header of ipc message from %s failed\n",
+				ipc_msg_file);
+			exit(EXIT_FAILURE);
+		}
+
+		fprintf(stdout, "header read: %ud\n", fuzzer.msg.header);
+
+		unsigned int uret = 0;
+		uret = fread(&fuzzer.msg.msg_size, sizeof(unsigned int), 1, fp);
+
+		fprintf(stdout, "msg_size read: %lu\n", fuzzer.msg.msg_size);
+		if (uret != 1) {
+			fprintf(stderr, "error: reading msg_size of ipc message from %s failed\n",
+				ipc_msg_file);
+			exit(EXIT_FAILURE);
+		}
+
+		if (fuzzer.msg.msg_size > SOF_IPC_MSG_MAX_SIZE) {
+			fprintf(stderr, "error: msg_size of ipc message exceeds max size %d\n",
```

Comment:
I am not sure we want to put this restriction on the fuzzer, this might be something we want to let it fuzz.

Reply:
Yeah. You are right here. Let's do that.